src/utils/hooks/useEmojiAutocomplete.ts

```diff
@@ -0,0 +1,215 @@
+import {ChangeEventHandler, FormEventHandler, HTMLProps, KeyboardEventHandler, useCallback, useEffect, useState, ComponentProps, ReactEventHandler} from "react";
+import {MAX_NOTE_LENGTH, MIN_CHARACTERS_TO_TRIGGER_EMOJI_SUGGESTIONS} from "constants/misc";
+import {EmojiSuggestions} from "components/EmojiSuggestions";
+import {SkinToneComponent} from "types/skinTone";
+import {useAppSelector} from "store";
+import {useOnBlur} from "./useOnBlur";
+
+export const emojiRegex = /^:([\w\d]+):?$/i;
+
+export type EmojiData = [slug: string, emoji: string, supportsSkintones: boolean, names: string[]];
+
+type InputElement = HTMLTextAreaElement | HTMLInputElement;
+
+export function emojiWithSkinTone(emoji: string, skinTone: SkinToneComponent) {
+  // the emoji with skin color support is assumed to be the first unicode character
+  // this means emojis with multiple people are unfortunately not supported
+  return emoji.substring(0, 2) + skinTone + emoji.substring(2);
+}
+
+/**
+ * this hook handles all the logic for the emoji autocomplete
+ * if you want to extend the event handlers, first call the handler retured,
+ * then check with e.defaultPrevented if the input has something to do with the emoji suggestions
+ *
+ * @param maxInputLength the maximum length of the input
+ * @param initialValue the initial value of the input
+ * @param suggestionsHidden whether the suggestions should be hidden (affects keyboard navigation)
+ * @returns `containerRef`: set this to the container element that includes both the input and the `EmojiSuggestions` component (used for onBlur)
+ * @returns `suggestionsProps`: props for the EmojiSuggestions component
+ * @returns `value` and `setValue`: get and set the value of the input
+ * @returns `inputBindings`: bindings for the input element (onChange, onKeyDown, onKeyUp, onClick, maxLength, value)
+ */
+export const useEmojiAutocomplete = <ContainerElement extends HTMLElement>(
+  {
+    maxInputLength = MAX_NOTE_LENGTH,
+    initialValue = "",
+    suggestionsHidden = false,
+  }: {
+    maxInputLength?: number;
+    initialValue?: string;
+    suggestionsHidden?: boolean;
```

Comment:
```suggestion
    maxInputLength: number;
    initialValue: string;
    suggestionsHidden: boolean;
```

Since you provide default values above, theses props are always provided. Or am I missing something?

Reply:
If I'd leave out the `?` it would require the properties to allways be specified/present in the object when calling the function. I added them so you can specify only what you need

```ts
// currently possible, but impossible with "?" removed
useEmojiAutocomplete({
  initialValue: "lorem ipsum"
})
```